File: lib/core/utils/url_normalizer.dart
String normalizeUrl(String url) {
  var normalized = url.trim();
  if (normalized.isEmpty) return '';
  if (!normalized.startsWith('http://') && !normalized.startsWith('https://')) {
    normalized = 'http://$normalized';
  }
  normalized = normalized.replaceAll(RegExp(r'/+$'), '');
  return normalized;
}

/// Oculta usuario y contraseña en URLs Xtream directas
/// (`/live/user/pass/id.ts`) y en parámetros `username`/`password`, antes de
/// mostrar texto o registrarlo.
String sanitizeStreamUrl(String text) {
  final withoutQueryCredentials = text.replaceAllMapped(
    RegExp(
      "([?&](?:username|password)=)[^&\\s'\"]+",
      caseSensitive: false,
    ),
    (match) => '${match.group(1)}***',
  );
  return withoutQueryCredentials.replaceAllMapped(
    RegExp("((?:live|movie|series|timeshift)/)[^/]+/[^/]+(/[^'\"\\s]*)"),
    (match) => '${match.group(1)}***/***${match.group(2)}',
  );
}

File: app/src/main/java/com/algoce95/novaiptv/core/utils/UrlNormalizer.kt
package com.algoce95.novaiptv.core.utils

/** Paridad con `lib/core/utils/url_normalizer.dart` de Flutter. */
object UrlNormalizer {
    fun normalizeUrl(url: String): String {
        var normalized = url.trim()
        if (normalized.isEmpty()) return ""
        if (!normalized.startsWith("http://") && !normalized.startsWith("https://")) {
            normalized = "http://$normalized"
        }
        return normalized.trimEnd('/')
    }

    private val queryCredentials =
        Regex("([?&](?:username|password)=)[^&\\s'\"]+", RegexOption.IGNORE_CASE)
    private val pathCredentials =
        Regex("((?:live|movie|series|timeshift)/)[^/]+/[^/]+(/[^'\"\\s]*)")

    /** Oculta usuario y contraseña antes de mostrar texto o registrarlo. */
    fun sanitizeStreamUrl(text: String): String {
        val withoutQuery = queryCredentials.replace(text) { "${it.groupValues[1]}***" }
        return pathCredentials.replace(withoutQuery) {
            "${it.groupValues[1]}***/***${it.groupValues[2]}"
        }
    }
}
